File: databricks_notebooks/python/no_mlflow/01_Write_Bundle.py
# Databricks notebook source
# MAGIC %md ## Write model as MLeap bundle to filesystem

# COMMAND ----------

# MAGIC %md ### Setup

# COMMAND ----------

# MAGIC %run ../Common

# COMMAND ----------

# MAGIC %md ### Read data

# COMMAND ----------

data = spark.read.csv(data_path, header="true", inferSchema="true")
(trainingData, testData) = data.randomSplit([0.7, 0.3], 2019)

# COMMAND ----------

# MAGIC %md ### Train pipeline

# COMMAND ----------

from pyspark.ml import Pipeline
from pyspark.ml.regression import DecisionTreeRegressor
from pyspark.ml.feature import VectorAssembler

maxDepth = 16
dt = DecisionTreeRegressor(labelCol="quality", featuresCol="features", maxDepth=maxDepth)
assembler = VectorAssembler(inputCols=data.columns[:-1], outputCol="features")
pipeline = Pipeline(stages=[assembler, dt])

model = pipeline.fit(trainingData)
predictions = model.transform(testData)

# COMMAND ----------

display(predictions)

# COMMAND ----------

# MAGIC %md ### Write MLeap bundle

# COMMAND ----------

if os.path.exists(bundle_path_python):
    os.remove(bundle_path_python)
bundle_path_python

# COMMAND ----------

from mleap.pyspark.spark_support import SimpleSparkSerializer
model.serializeToBundle("jar:file:"+bundle_path_python, predictions)

# COMMAND ----------

# MAGIC %md ### Write data schema file

# COMMAND ----------

schema_path = os.path.join(work_dir_python,"wine-schema.json")
with open(schema_path, "w") as f:
    f.write(data.schema.json())
schema_path

# COMMAND ----------

# MAGIC %md ### Check working files

# COMMAND ----------

# MAGIC %sh echo $WORK_DIR

# COMMAND ----------

# MAGIC %sh ls -l $WORK_DIR/python

File: databricks_notebooks/scala/no_mlflow/01_Write_Bundle.scala
// Databricks notebook source
// MAGIC %md ## Write model as MLeap bundle to filesystem

// COMMAND ----------

// MAGIC %md ### Setup

// COMMAND ----------

// MAGIC %run ../Common

// COMMAND ----------

// MAGIC %run ../SparkBundleUtils

// COMMAND ----------

// MAGIC %md ### Read data

// COMMAND ----------

val data = spark.read.format("csv")
  .option("header", "true")
  .option("inferSchema", "true")
  .load(dataPathDbfs)
val Array(trainingData, testData) = data.randomSplit(Array(0.7, 0.3), 2019)

// COMMAND ----------

// MAGIC %md ### Train Pipeline

// COMMAND ----------

import org.apache.spark.ml.Pipeline
import org.apache.spark.ml.feature.VectorAssembler
import org.apache.spark.ml.regression.DecisionTreeRegressor

// COMMAND ----------

// Setup
val maxDepth = 16
val (colLabel, colPrediction, colFeatures) = ("quality", "prediction", "features")

// Create model
val dt = new DecisionTreeRegressor()
  .setLabelCol(colLabel)
  .setFeaturesCol(colFeatures)
  .setMaxDepth(maxDepth)

// Create pipeline
val columns = data.columns.toList.filter(_ != colLabel)
val assembler = new VectorAssembler()
  .setInputCols(columns.toArray)
  .setOutputCol(colFeatures)
val pipeline = new Pipeline().setStages(Array(assembler,dt))

// Fit model
val model = pipeline.fit(trainingData)

// Transform
val predictions = model.transform(testData)

// COMMAND ----------

// MAGIC %md ### Write model as MLeap bundle

// COMMAND ----------

val bundleUri = "jar:file:" + bundlePathSpark
SparkBundleUtils.writeModel(bundleUri, model, predictions)

// COMMAND ----------

// MAGIC %md ### Write data schema file

// COMMAND ----------

val schemaPath = workDirScala + "/wine-schema.json"
new java.io.PrintWriter(schemaPath) { write(data.schema.json) ; close }

// COMMAND ----------

// MAGIC %md ### Check working files

// COMMAND ----------

// MAGIC %sh ls -l /dbfs/tmp/andre.mesarovic@databricks.com/mleap_demo/scala